client-misk/src/test/kotlin/app/cash/backfila/client/misk/hibernate/VitessShardedBackfillTest.kt

```diff
@@ -0,0 +1,319 @@
+    assertThat(run.backfill.idsRanDry).isEmpty()
+    assertThat(run.backfill.idsRanWet).doesNotContainAnyElementsOf(expectedIds)
+  }
+
+  private fun checkRightShardDoesNothing(run: BackfillRun<TestBackfill>) {
+    // Right Shard will not match ot backfill anything since chickfila is on the left shard
+    while (!run.instanceProgressSnapshot[RIGHT_SHARD].done) {
+      val rightScan = run.shardScan(RIGHT_SHARD)
+      if ( rightScan.batches.isNotEmpty()) {
+        assertThat(rightScan.batches.single().matching_record_count).isEqualTo(0)
+      }
+      assertThat(run.backfill.idsRanDry).isEmpty()
+      assertThat(run.backfill.idsRanWet).isEmpty()
+    }
+    assertThat(run.instanceProgressSnapshot[RIGHT_SHARD]).isDone()
+  }
+
+  private fun BackfillRun<*>.configureForTest() {
+    this.batchSize = 10L
+    this.scanSize = 100L
+    this.computeCountLimit = 1L
+  }
+
+  private fun createSomeOrders(): List<Id<DbOrder>> {
+    createRestaurants()
+    return transacter.allowCowrites().transaction { session: Session ->
+      val expected = mutableListOf<Id<DbOrder>>()
+      repeat((1..10).count()) {
+        val id = session.save(DbOrder(CHICKFILA))
+        expected.add(id)
+      }
+
+      // Intersperse these to make sure we test skipping non matching records.
+      repeat((1..5).count()) { session.save(DbOrder(MCDONALDS)) }
+      repeat((1..2).count()) { session.save(DbOrder(WENDYS)) } // On the other shard
+
+      repeat((1..10).count()) {
+        val id = session.save(DbOrder(CHICKFILA))
+        expected.add(id)
+      }
+      expected.toImmutableList()
+    }
+  }
+
+  private fun createNoMatchingOrders() {
+    createRestaurants()
+    transacter.allowCowrites().transaction { session: Session ->
+      repeat((1..5).count()) { session.save(DbOrder(MCDONALDS)) }
+      repeat((1..2).count()) { session.save(DbOrder(WENDYS)) } // On the other shard
+    }
+  }
+
+  private fun createRestaurants() {
+    transacter.allowCowrites().transaction { session: Session ->
+      val chickfila = session.save(DbRestaurant(CHICKFILA,"chickfila"))
+      assertThat(chickfila).isEqualTo(CHICKFILA)
+      assertThat(chickfila.shard(session)).isEqualTo(LEFT_SHARD)
+      // Makes sure some non-matching make it on each shard, hash results should be a stable assumption.
+      val mcdonalds = session.save(DbRestaurant(MCDONALDS, "mcdonalds"))
+      assertThat(mcdonalds).isEqualTo(MCDONALDS)
+      assertThat(mcdonalds.shard(session)).isEqualTo(LEFT_SHARD)
+      val arbys = session.save(DbRestaurant(ARBYS, "arbys"))
+      assertThat(arbys).isEqualTo(ARBYS)
+      assertThat(arbys.shard(session)).isEqualTo(LEFT_SHARD)
+      val wendys = session.save(DbRestaurant(WENDYS, "wendys"))
+      assertThat(wendys).isEqualTo(WENDYS)
+      assertThat(wendys.shard(session)).isEqualTo(RIGHT_SHARD)
+    }
+  }
+
+  companion object {
+    val KEYSPACE = DbRestaurant::class.java.getAnnotation(misk.hibernate.annotation.Keyspace::class.java).keyspace()
```

Comment:
why is `Keyspace` fully qualifed?

Reply:
There is the Keyspace annotation and the class. They are different. Funnily KEYSPACE has the type Keyspace but the class.